text/0003-hash-indexed-object-store.md

```diff
@@ -0,0 +1,624 @@
+
+## Summary
+
+[summary]: #summary
+
+RFC PR: [casperlabs/rfcs#0003](https://github.com/casperlabs/rfcs/pull/0003)
+
+There is a variety of "objects" floating around in our codebase at the moment: Deploys, WASM blobs, Blocks, DAG nodes and others. By standardizing them into a [Git](https://git-scm.com)-inspired object model we can simplify our networking, gossiping, retrieval, and storage logic greatly.
+
+## Motivation
+
+[motivation]: #motivation
+
+One possible implementation of a mechanism for a new node joining the network is to provide it with a block and have it download all parents and dependencies up until the well-known genesis block.
+
+With the current node architecture, this approach would require quite a few component instances: Separate instantiations of the gossiper and fetcher components for each type involved, which at the minimum are block and deploy if not counting DAG nodes. Storage needs to namespace according to types and component developers need to remember which effect to call for each. During all of this, the core logic remains the same for all IO operations of these different types though.
+
+By unifying all these types, we can simplify all of our IO components drastically. Ensuring that all have a unique hash that includes a type-discriminating tag ensures that we can use the same namespace for identifiers.
+
+As a result, our IO layer will be greatly simplified, as well as its usage. Possible gains are easier snapshotting, DB handling and shipping data around in general. Speed-ups through external caching become very real (see [Future possibilities](#future-possibilities)).
+
+Taking this one step further, we can express the tree structure given through hash references independent of the objects type as well, moving more IO logic away from business components.
+
+## Guide-level explanation
+
+[guide-level-explanation]: #guide-level-explanation
+
+(It is recommended to read the [prior art](#prior-art) section beforehand)
+
+Every *object* that our node handles has common properties, namely:
+
+* It can be serialized to binary storage.
+* It has a unique identifying hash that depends solely on its contents.
+* It may reference other objects.
+
+This is common to everything we handle, for example a block will reference a number of deploys, which in turn reference WASM blobs.
+
+### Serialization
+
+Every object can be serialized to be sent across the network, which also means it can be stored in storage. In general, any serialized object is prefixed with a short tag indicating its type, regardless of the chosen serialization format.
+
+This helps disamgiuate between different objects whose serialization may be the same under a given format, and can also be used to avoid conflicts with versioning, should an existing storage be run with a different node version.
+
+The tag should be kept short, one byte should be plenty to cover all of our usecases (we will reserve one value for future expansion if necessary).
```

Comment:
```suggestion
A type tag helps disambiguate objects whose serialization may be the same under a given format, and can also be used to avoid conflicts across different node versions. One byte should be sufficient to cover all of our use cases (one value could be reserved for future expansion if necessary).
```